### PR TITLE
GITC-557/556: claim history + ready to claim

### PR DESCRIPTION
**Pull request by chibie**
##### Description

TODO:

- [x] Empty state when user has no grant
- [x] Fetch all grants of a user not just only grants with clr matches
- [x] fix sticky notes z-index issue
- [x] ensure user is not on polygon network
- [x] ensure user is in right payout network
- [x] sort grants based on existing matching data

##### CLAIM HISTORY

- [x] GIVEN I am a grant owner who has a claim history
WHEN I go to look up my claim history
THEN I will navigate to the new "matching funds" page.

- [x] GIVEN I am a grant owner who has a claim history
WHEN I go to look up my claim history on the new "matching funds" page.
THEN I will navigate to the "claim history" tab.

- [x] GIVEN I am a grant owner who has a claim history
WHEN I am looking at the new "matching funds" page under the "claim history" tab
THEN I will see matching pool claim details (title, amount, status, Etherscan) by grant.

- [x] GIVEN I am looking at matching pool claim details by grant on the new "matching funds" page under the "claim history" tab
WHEN I look at the matching pool claim details for one specific grant
THEN I see two of three possible statuses: previously claimed a match amount "claimed" OR has previously clawed back match funds "returned to matching pool".

- [x] GIVEN I am viewing the "claim history" tab
WHEN I scroll down on the page
THEN I will still be able to see the “Ready to Claim” and “Claim History” tabs even if I’m not scrolled all the way to the top.

##### READY TO CLAIM

- [x] GIVEN I am looking at matching pool claim details by grant on the new "matching funds" page under the "ready to claim" tab
AND the claim period is active
WHEN I look at the matching pool claim details for one specific grant
THEN I see the status "ready to claim".

- [x] GIVEN I am looking at matching pool claim details by grant on the new "matching funds" page under the "ready to claim" tab
AND the claim period is active
WHEN I look at the matching pool claim details for one specific grant
THEN the claim button is clickable.

- [x] GIVEN I am viewing the "ready to claim" tab
WHEN I scroll down on the page
THEN I will still be able to see the “Ready to Claim” and “Claim History” tabs even if I’m not scrolled all the way to the top.

- [x] GIVEN I am looking at matching pool claim details by grant on the new "matching funds" page under the "ready to claim" tab during an active claim period
WHEN I click "claim"
THEN metamask is prompted to sign the transaction.

- [x] GIVEN I have signed the metamask transaction from hitting "claim"
WHEN the transaction is processing
THEN I see the matching funds page "claim history" tab with the default status "processing" on my matching claim transaction.

- [x] GIVEN I have signed the metamask transaction from hitting "claim"
WHEN the transaction is successful
THEN I see the matching funds page "claim history" tab with the status "claimed" on my matching claim transaction.

- [x] GIVEN I have signed the metamask transaction from hitting "claim"
WHEN the transaction fails
THEN the user should be redirected back to the "ready to claim" tab.

##### NOTE

https://gist.github.com/thelostone-mc/d0781cb28845fbb32917c794d5f2068a has to be run once migrations are run in live to populate claim history

### CODE REVIEW

File: app/grants/models/grant.py [hunk 1]
```diff
@@ -24,6 +24,37 @@
...
+    )
+    funding_withdrawn = models.BooleanField(
+        default=False,
+        help_text=_('Was the Matching Contract funding withdrawn?')
+    )
+    funding_withdrawal_date = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text=_('When was funding in Matching Contract withdrawn?')
+    )
```
Comment (thelostone-mc) [inline app/grants/models/grant.py:52]: As opposed to having 

- funding_withdrawn
- ready_to_claim


Why not instead have a single field ? https://github.com/gitcoinco/web/pull/9657/files#r768346087

File: app/grants/models/grant.py [hunk 2]
```diff
@@ -24,6 +24,37 @@
...
+    payout_token = models.CharField(
+        max_length=10,
+        default='DAI',
+        help_text=_('Currency in which funds would be paid')
+    )
+    funding_withdrawn = models.BooleanField(
+        default=False,
+        help_text=_('Was the Matching Contract funding withdrawn?')
+    )
+    funding_withdrawal_date = models.DateTimeField(
```
Comment (thelostone-mc) [inline app/grants/models/grant.py:48]: Why is this needed?

**Review by Mehdish1414 — APPROVED** [review]

File: app/grants/models/grant.py (hunk 2 above)
Comment (chibie) [reply app/grants/models/grant.py:48]: @thelostone-mc the design has a date beneath a clawed back status

File: app/grants/models/grant.py (hunk 1 above)
Comment (chibie) [reply app/grants/models/grant.py:52]: oh yes! i'll change this @thelostone-mc

**Comment by thelostone-mc** [comment]
@chibie could we add this field to GrantPayout as well (cause we might need this in the next round) and having this created before hand -> makes it easier to navigate that

```

+    NETWORKS = [
+        ('mainnet', 'mainnet'),
+        ('rinkeby', 'rinkeby'),
+        # ('polygon', 'polygon'),
+        # ('mumbai', 'mumbai'),
+    ]


+    network = models.CharField(
+        max_length=15,
+        default='mainnet',
+        choices=NETWORKS,
+        help_text=_('Network where contract is deployed')
+    )
```

File: app/assets/v2/js/grants/matching_funds.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+      const txHash = match.claim_tx;
+
+      let status = 'not-found';
+      let timestamp = null;
+
+      if (!txHash) {
+        return { status, timestamp };
+      }
+
+      web3 = new Web3(`wss://mainnet.infura.io/ws/v3/${document.contxt.INFURA_V3_PROJECT_ID}`);
```
Comment (thelostone-mc) [inline app/assets/v2/js/grants/matching_funds.js:96]: let's not hardcode mainnet. let's wrap it into a function where we can pass the network and get the web3. we might want to use rinkeby / matic / etc

File: app/assets/v2/js/grants/matching_funds.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+      waitingState(true);
+
+      // Connect wallet
+      if (!provider) {
+        await onConnect();
+      }
+
+      // Confirm wallet was connected (user may have closed wallet connection prompt)
+      if (!provider) {
+        return;
```
Comment (thelostone-mc) [inline app/assets/v2/js/grants/matching_funds.js:143]: maybe add an alert here

**Comment by thelostone-mc** [comment]
**PENDING:** 

- payout_round_noncustodial.py 
   - L149 has to be updated to ensure CLRMatch's `grant_payout` is updated
   - L102 -> provdier is harcoded to mainnet / rinkeby

File: app/grants/models/grant.py [hunk 5]
```diff
@@ -18,12 +18,69 @@
...
+    PAYOUT_STATUS = [
+        ('pending', 'pending'),
+        ('ready', 'ready'),
+        ('expired', 'expired'),
+        ('funding_withdrawn', 'funding_withdrawn')
+    ]
+    NETWORKS = [
+        ('mainnet', 'mainnet'),
+        ('rinkeby', 'rinkeby'),
+        # ('polygon', 'polygon'),
```
Comment (thelostone-mc) [inline app/grants/models/grant.py:38]: this is commented out for now as we support only mainnet for this round payout

File: app/grants/models/grant.py [hunk 6]
```diff
@@ -18,12 +18,69 @@
...
+        default='mainnet',
+        choices=NETWORKS,
+        help_text=_('Network where contract is deployed')
+    )
+    payout_token = models.CharField(
+        max_length=10,
+        default='DAI',
+        help_text=_('Currency in which funds would be paid')
+    )
+    # payout_token_address = models.CharField(
```
Comment (thelostone-mc) [inline app/grants/models/grant.py:63]: this is commented out for now due as we support only DAI payouts but that will change post this round

**Comment by thelostone-mc** [comment]
@chibie could we add in tests for the model changes ?

**Review by nutrina — APPROVED** [review]
This looks good to me.

**Comment by frankchen07** [comment]
just looking on staging, although not sure if it's ready for staging - why does visiting nate's grant and clicking on claim funds lead to a page with fronk's grants?:

Screen Shot 2022-01-07 at 12 23 04
Screen Shot 2022-01-07 at 12 23 11
Screen Shot 2022-01-07 at 12 23 16

**Review by thelostone-mc — APPROVED** [review]

**Comment by nfrgosselin** [comment]
This looks good to me @chibie and @thelostone-mc!

I just want @frankchen07 to give a final okay before we call this approved from the product side.

**Comment by frankchen07** [comment]
Just one comment from me from two of the A/C criteria:

> GIVEN I am looking at matching pool claim details by grant on the new "matching funds" page under the "claim history" tab
> WHEN I look at the matching pool claim details for one specific grant
> THEN I see two of three possible statuses: previously claimed a match amount "claimed" OR has previously clawed back match funds "returned to matching pool".

^ I don't see a status on the Claim History page. 

Screen Shot 2022-01-10 at 11 47 43

> GIVEN I have signed the metamask transaction from hitting "claim"
> WHEN the transaction is processing
> THEN I see the matching funds page "claim history" tab with the default status "processing" on my matching claim transaction.

^ I don't think I saw "processing" unless 1. I just missed it, or 2. we're unable to grab that information.

===

Otherwise, LGTM, all other A/C criteria look to be fulfilled.
